src/pubsub.js

```diff
@@ -0,0 +1,508 @@
+          .then((subscription) => {
+            s = subscription
+            return ipfs.pubsub.peers(topicName)
+          })
+          .then((peers) => {
+            expect(peers).to.exist
+            s.cancel(done)
+          })
+          .catch(done)
+      })
+
+      it('topics', (done) => {
+        ipfs.pubsub.ls()
+          .then((topics) => {
+            expect(topics).to.exist
+            expect(topics.length).to.equal(0)
+            done()
+          })
+          .catch(done)
+      })
+    })
+
+    describe('load tests', () => {
+      it('send/receive 10k messages', (done) => {
+        const expectedString = 'hello'
+        const count = 10000
+        let sendCount = 0
+        let receivedCount = 0
+        let startTime
+
+        ipfs.pubsub.subscribe(topicName, (err, subscription) => {
+          expect(err).to.not.exists
+
+          const outputProgress = () => {
+            process.stdout.write('                                                                        \r')
+            process.stdout.write('Sent: ' + sendCount + ' of ' + count + ', Received: ' + receivedCount + '\r')
+          }
+
+          subscription.on('data', (d) => {
+            expect(d.data).to.be.equal(expectedString)
+            receivedCount++
+            outputProgress()
+            if (receivedCount >= count) {
+              const duration = new Date().getTime() - startTime
+              process.stdout.write('                                                                        \r')
+              console.log(`Send/Receive 10k messages took: ${duration} ms, ${Math.floor(count / (duration / 1000))} ops / s`)
+              subscription.cancel(done)
+            }
+          })
+
+          const loop = () => {
+            if (sendCount < count) {
+              sendCount++
+              outputProgress()
+              ipfs2.pubsub.publish(topicName, expectedString, (err) => {
+                expect(err).to.not.exist
+                process.nextTick(() => loop())
+              })
+            }
+          }
+
+          waitForPeers(ipfs, [ipfs2.PeerId], (err) => {
+            expect(err).to.not.exist
+            startTime = new Date().getTime()
+            loop()
+          })
+        })
+      })
+
+      it('call publish 1k times', (done) => {
+        const expectedString = 'hello'
+        const count = 1000
+        let sendCount = 0
+
+        const loop = () => {
+          if (sendCount < count) {
+            sendCount++
+            process.stdout.write('                                               \r')
+            process.stdout.write('Sending messages: ' + sendCount + ' of ' + count + '\r')
+            ipfs.pubsub.publish(topicName, expectedString, (err) => {
+              expect(err).to.not.exist
+              process.nextTick(() => loop())
+            })
+          } else {
+            done()
+          }
+        }
+        loop()
+      })
+
+      it('call subscribe 1k times', (done) => {
```

Comment:
Unsubscribe changes the internal state. This test is meant to make sure that NOT calling unsubscribe before calling subscribe again doesn't change the state and works as expected.